src/System.Private.CoreLib/shared/System/Threading/ThreadPool.cs

```diff
@@ -48,413 +46,1245 @@ internal static class ThreadPoolGlobals
+                                    if (sequenceNumber == position)
+                                    {
+                                        // update Enqueue - must do before marking the slot full.
+                                        // otherwise someone could lock the full slot while having stale Enqueue.
+                                        _queueEnds.Enqueue = position + 1;
+                                        slot.Item = item;
+
+                                        // make the slot appear full in the current generation.
+                                        // since the slot on the left is still locked, only poppers/enqueuers can use it, but can use immediately
+                                        Volatile.Write(ref slot.SequenceNumber, position + Full);
+
+                                        // unlock prev slot
+                                        // must be after we moved enq to the next slot, or someone may pop prev and break continuity of full slots.
+                                        prevSlot.SequenceNumber = prevSequenceNumber;
+                                        return true;
+                                    }
+
+                                    // do we see the prev generation?
+                                    if (position - sequenceNumber > 0)
+                                    {
+                                        // Set Enqueue to throw off anyone else trying to enqueue or pop, unless we have already done that.
+                                        // we need a fence between writing to Enqueue and unlocking, but we unlock with a CAS anyways
+                                        _queueEnds.Enqueue = position + FreezeOffset;
+                                        _frozenForEnqueues = true;
+                                    }
+                                }
+
+                                // enqueue changed or segment is full (rare cases)
+                                // unlock the slot through CAS in case slot was robbed
+                                Interlocked.CompareExchange(ref prevSlot.SequenceNumber, prevSequenceNumber, prevSequenceNumber + Change);
                             }
                         }
-                        finally
+
+                        if (_frozenForEnqueues)
                         {
-                            if (lockTaken)
-                                m_foreignLock.Exit(useMemoryBarrier: false);
+                            return false;
                         }
+
+                        // Lost a race. Most likely to the dequeuer of the last remaining item, which will be gone shortly. Try again.
+                        // NOTE: We only need a compiler fence here. As long as we re-read Enqueue, it could be an ordinary read.
+                        //       This is not a common code path though, so volatile read will do.
+                        position = Volatile.Read(ref _queueEnds.Enqueue);
```

Comment:
There is a good possibility that this path would busy-spin due to old information read at the beginning. That would not change with the `Volatile.Read`, though I see the need for that. Any busy-spin paths should spin-wait properly (including a `Thread.SpinWait` and `Thread.Sleep(0)` somewhere), as busy-spinning can take away time from other threads that may do more meaningful work, especially on machines that don't have hyperthreading, but even in those cases.

The same applies to many other cases in this change where spin-waiting is avoided in favor of busy-spinning. Busy-spinning is somewhat ok only when one thread is guaranteed to make progress, such as a path where there is an interlocked operation that is guaranteed to make progress on one thread. When there is a path that is not guaranteed to make progress (decision about contention is made based on unsynchronized memory reads), that can turn into a potentially bad busy-spin and how bad it would be would depend on the scheduling policy.